# Merchant Ranking System

This notebook develops an interpretable decision-support heuristic for BNPL merchant onboarding, not a statistically optimal prediction model.

It combines merchant value, customer reach/scale, growth, stability and market context into a business score, then applies a 10% fraud-safety adjustment using Member 4's consumer exposure and KNN-estimated merchant risk signals. KNN risk estimates are relative risk signals, not precise fraud probabilities.

In [1]:
from pathlib import Path
import pandas as pd
import numpy as np

# Support execution from the repository root or the notebook directory.
cwd = Path.cwd().resolve()
PROJECT_ROOT = next(
    (path for path in (cwd, *cwd.parents)
     if (path / "member3_merchant_features").is_dir()
     and (path / "member4_fraud").is_dir()),
    None,
)
if PROJECT_ROOT is None:
    raise FileNotFoundError("Could not locate the repository root.")
RESULTS_DIR = PROJECT_ROOT / "member5_ranking" / "results"
RESULTS_DIR.mkdir(parents=True, exist_ok=True)

INPUT_PATH = (
    PROJECT_ROOT
    / "member3_merchant_features"
    / "results"
    / "merchant_features.parquet"
)

df = pd.read_parquet(INPUT_PATH)

print("Shape:", df.shape)
df.head()

Shape: (4422, 46)


,merchant_abn,merchant_name,merchant_category,merchant_pricing_level,merchant_take_rate_pct,has_merchant_master_record,total_transactions,total_revenue,avg_transaction_value,unique_consumers,...,census_single_parent_family_share,census_couple_with_children_family_share,census_unemployment_rate_published_pct,census_labour_force_participation_rate_published_pct,seifa_irsd_national_decile,seifa_irsad_national_decile,seifa_ier_national_decile,ato_taxable_income_or_loss_per_reporter,ato_salary_or_wages_per_recipient,ato_net_tax_payer_share
0,10023283211,Felis Limited,"furniture, home furnishings and equipment shop...",E,0.18,True,3261,703277.711451,215.663205,3032,...,0.137732,0.396714,4.545196,59.159350,5.426345,5.412987,5.455659,68260.371739,63998.111541,0.734878
1,10142254217,Arcu Ac Orci Corporation,"cable, satellite, and other pay television and...",B,4.22,True,3036,118356.146073,38.984238,2849,...,0.137717,0.397312,4.618711,59.135731,5.512892,5.524395,5.500595,69361.310505,64617.995761,0.734825
2,10165489824,Nunc Sed Company,"jewelry, watch, clock, and silverware shops",B,4.40,True,5,56180.473857,11236.094771,5,...,0.114423,0.490410,4.150000,63.625000,8.250000,8.000000,8.500000,76144.193842,63913.702304,0.782871
3,10187291046,Ultricies Dignissim Lacus Foundation,"watch, clock, and jewelry repair shops",B,3.29,True,336,39693.730387,118.136102,335,...,0.132108,0.393778,4.485274,59.633219,5.513793,5.434483,5.548276,68884.847450,65123.944932,0.734622
4,10192359162,Enim Condimentum PC,"music shops - musical instruments, pianos, and...",A,6.33,True,385,177980.505456,462.287027,383,...,0.136463,0.401961,4.413313,60.091331,5.744548,5.679128,5.732087,70416.453471,66239.804559,0.746924


## 1. Candidate Pool

The feature table contains 4,422 merchants; 4,026 have valid merchant master records and are eligible.
The 396 transaction-only merchants lack category and take-rate data required for comparable scoring.
They are excluded for missing information, not because they have been shown to be poor onboarding choices.

In [2]:
eligible_df = df[df["has_merchant_master_record"]].copy()

print("All merchants:", len(df))
print("Eligible merchants:", len(eligible_df))
print("Excluded orphan merchants:", len(df) - len(eligible_df))

All merchants: 4422
Eligible merchants: 4026
Excluded orphan merchants: 396


## 2. Candidate Ranking Features

The business-only stage considers value, customer reach/scale, historical growth, stability and
regional context. CustomerScore uses `unique_consumers`; repeat share and transaction count are
diagnostics only. Fraud is already integrated in sections 8-10, using consumer exposure and KNN
merchant risk. The legacy `preliminary_*` outputs denote the business-only stage, not unfinished work.

In [3]:
candidate_features = [
    "estimated_bnpl_revenue",
    "total_revenue",
    "total_transactions",
    "unique_consumers",
    "avg_transaction_value",
    "repeat_consumer_share",
    "normalized_monthly_revenue_trend",
    "monthly_revenue_cv",
    "low_sample_growth_estimate",
    "regional_data_coverage_rate_all_sources_by_count",
    "census_median_household_income_weekly",
    "seifa_irsad_national_decile",
    "ato_taxable_income_or_loss_per_reporter",
]

eligible_df[candidate_features].describe().T

,count,mean,std,min,25%,50%,75%,max
estimated_bnpl_revenue,4026.0,23933.339418,5.849947e+04,12.313488,1470.210886,5146.420842,22933.624931,6.638703e+05
total_revenue,4026.0,535781.369254,1.218831e+06,10064.933919,36652.804201,141397.947807,628006.258580,9.857402e+06
total_transactions,4026.0,3381.687779,1.414670e+04,1.000000,93.000000,418.000000,2057.500000,2.895130e+05
unique_consumers,4026.0,2025.218579,3.875050e+03,1.000000,93.000000,415.500000,1961.000000,2.408100e+04
avg_transaction_value,4026.0,1159.595793,2.947518e+03,7.588294,118.099522,317.329914,836.237711,5.187664e+04
repeat_consumer_share,4026.0,0.051295,1.185795e-01,0.000000,0.000000,0.009718,0.041562,9.999169e-01
normalized_monthly_revenue_trend,4019.0,0.011004,4.624178e-02,-0.857143,0.006984,0.014355,0.020337,6.551908e-01
monthly_revenue_cv,4019.0,0.482603,4.134482e-01,0.133836,0.217327,0.321069,0.577596,3.872983e+00
regional_data_coverage_rate_all_sources_by_count,4026.0,0.806718,5.417902e-02,0.000000,0.796875,0.807871,0.819749,1.000000e+00
census_median_household_income_weekly,4026.0,1603.516538,8.298170e+01,806.500000,1587.309716,1604.879927,1620.820699,3.124000e+03


In [4]:
feature_missingness = (
    eligible_df[candidate_features]
    .isna()
    .mean()
    .sort_values(ascending=False)
    .to_frame("missing_rate")
)

feature_missingness

,missing_rate
normalized_monthly_revenue_trend,0.001739
monthly_revenue_cv,0.001739
ato_taxable_income_or_loss_per_reporter,0.000248
estimated_bnpl_revenue,0.000000
total_revenue,0.000000
total_transactions,0.000000
unique_consumers,0.000000
avg_transaction_value,0.000000
repeat_consumer_share,0.000000
low_sample_growth_estimate,0.000000


## 3. Feature Redundancy Check

Before assigning weights, pairwise correlations are examined to reduce double-counting of closely related merchant characteristics.

In [5]:
correlation_features = [
    "estimated_bnpl_revenue",
    "total_revenue",
    "total_transactions",
    "unique_consumers",
    "avg_transaction_value",
    "repeat_consumer_share",
    "normalized_monthly_revenue_trend",
    "monthly_revenue_cv",
    "census_median_household_income_weekly",
    "seifa_irsad_national_decile",
    "ato_taxable_income_or_loss_per_reporter",
]

correlation_matrix = eligible_df[correlation_features].corr().round(3)

correlation_matrix.to_csv(
    PROJECT_ROOT / "member5_ranking" / "results" / "feature_correlation_matrix.csv"
)

correlation_matrix

,estimated_bnpl_revenue,total_revenue,total_transactions,unique_consumers,avg_transaction_value,repeat_consumer_share,normalized_monthly_revenue_trend,monthly_revenue_cv,census_median_household_income_weekly,seifa_irsad_national_decile,ato_taxable_income_or_loss_per_reporter
estimated_bnpl_revenue,1.000,0.923,0.555,0.613,-0.026,0.636,0.031,-0.239,0.007,0.003,0.004
total_revenue,0.923,1.000,0.569,0.648,-0.022,0.663,0.032,-0.254,0.007,0.003,0.003
total_transactions,0.555,0.569,1.000,0.714,-0.081,0.820,0.020,-0.164,0.005,0.003,0.003
unique_consumers,0.613,0.648,0.714,1.000,-0.168,0.977,0.042,-0.347,0.011,0.006,0.006
avg_transaction_value,-0.026,-0.022,-0.081,-0.168,1.000,-0.142,-0.214,0.546,-0.026,-0.000,-0.085
repeat_consumer_share,0.636,0.663,0.820,0.977,-0.142,1.000,0.035,-0.292,0.010,0.005,0.005
normalized_monthly_revenue_trend,0.031,0.032,0.020,0.042,-0.214,0.035,1.000,-0.223,0.216,0.141,0.132
monthly_revenue_cv,-0.239,-0.254,-0.164,-0.347,0.546,-0.292,-0.223,1.000,-0.045,-0.024,-0.028
census_median_household_income_weekly,0.007,0.007,0.005,0.011,-0.026,0.010,0.216,-0.045,1.000,0.751,0.654
seifa_irsad_national_decile,0.003,0.003,0.003,0.006,-0.000,0.005,0.141,-0.024,0.751,1.000,0.519


## 4. Selected Business Features

Highly correlated candidates are not all separately weighted, reducing avoidable double counting.
The selected inputs are:

| Dimension | Input | Direction / interpretation |
|---|---|---|
| Value | `estimated_bnpl_revenue` | Higher historical transaction value × take rate is preferred; not net profit. |
| Customer Reach | `unique_consumers` | More distinct historical customers; not repeat share or transaction count. |
| Growth | `normalized_monthly_revenue_trend` | Higher historical normalized slope; not a monthly growth percentage or forecast. |
| Stability | `monthly_revenue_cv` | Lower CV is preferred; reversed percentile. |
| Market Context | `seifa_irsad_national_decile` | Higher transaction-weighted customer-area advantage is the chosen contextual preference. |

Census and ATO remain contextual diagnostics, not directly scored. IRSAD does not measure individual
income, merchant location or proven profitability. The fraud adjustment is integrated separately below.

## 5. Percentile Normalisation

To combine features measured on different scales, each selected metric is converted to a percentile score from 0 to 100.

Higher percentile scores indicate stronger alignment with the chosen feature preferences.

For metrics where lower values are preferred, such as revenue volatility, the percentile score is reversed.

In [6]:
score_df = eligible_df.copy()

# Higher is better
score_df["value_score"] = (
    score_df["estimated_bnpl_revenue"]
    .rank(pct=True, method="average") * 100
)

score_df["customer_score"] = (
    score_df["unique_consumers"]
    .rank(pct=True, method="average") * 100
)

score_df["growth_score"] = (
    score_df["normalized_monthly_revenue_trend"]
    .rank(pct=True, method="average") * 100
)

score_df["market_score"] = (
    score_df["seifa_irsad_national_decile"]
    .rank(pct=True, method="average") * 100
)

# Lower volatility is better
score_df["stability_score"] = (
    100
    - score_df["monthly_revenue_cv"]
    .rank(pct=True, method="average") * 100
)

score_df[
    [
        "value_score",
        "customer_score",
        "growth_score",
        "stability_score",
        "market_score",
    ]
].describe().round(2)

,value_score,customer_score,growth_score,stability_score,market_score
count,4026.00,4026.00,4019.00,4019.00,4026.00
mean,50.01,50.01,50.01,49.99,50.01
std,28.87,28.87,28.87,28.87,28.87
min,0.02,0.10,0.02,0.00,0.02
25%,25.02,25.06,25.02,24.99,25.02
50%,50.01,50.02,50.01,49.99,50.01
75%,75.01,75.00,75.01,74.98,75.01
max,100.00,99.98,100.00,99.98,99.98


## 6. Preliminary Weighted Score

A preliminary balanced weighting scheme is used before fraud risk is integrated.

The current weights prioritise direct BNPL value and customer scale, while still allowing growth, stability and market context to influence the ranking.

In [7]:
# Neutral score for merchants whose growth/stability cannot be estimated
score_df["growth_score"] = score_df["growth_score"].fillna(50)
score_df["stability_score"] = score_df["stability_score"].fillna(50)

In [8]:
PRELIMINARY_WEIGHTS = {
    "value_score": 0.30,
    "customer_score": 0.25,
    "growth_score": 0.20,
    "stability_score": 0.15,
    "market_score": 0.10,
}

score_df["preliminary_score"] = sum(
    weight * score_df[column]
    for column, weight in PRELIMINARY_WEIGHTS.items()
)

score_df["preliminary_rank"] = (
    score_df["preliminary_score"]
    .rank(method="min", ascending=False)
    .astype(int)
)

preliminary_top_20 = (
    score_df
    .sort_values("preliminary_score", ascending=False)
    [
        [
            "preliminary_rank",
            "merchant_abn",
            "merchant_name",
            "merchant_category",
            "preliminary_score",
            "value_score",
            "customer_score",
            "growth_score",
            "stability_score",
            "market_score",
        ]
    ]
    .head(20)
)

preliminary_top_20

,preliminary_rank,merchant_abn,merchant_name,merchant_category,preliminary_score,value_score,customer_score,growth_score,stability_score,market_score
3966,1,90568944804,Diam Eu Dolor LLC,tent and awning shops,87.168863,99.230005,93.169399,75.167952,82.483205,67.014406
1414,2,38090089066,Interdum Feugiat Sed Inc.,"furniture, home furnishings and equipment shop...",86.800764,98.683557,99.254844,61.607365,93.928838,59.711873
869,3,27326652377,Tellus Aenean Corporation,"music shops - musical instruments, pianos, and...",86.253316,99.403875,89.195231,75.267479,86.165713,61.549925
3782,4,86772484982,Posuere Cubilia Curae LLC,"digital goods: books, movies, music",85.785250,95.032290,95.355191,59.442647,98.283155,68.057625
3692,5,84703983173,Amet Consulting,"computer programming , data processing, and in...",85.765124,96.572280,99.006458,68.847972,89.524757,48.435171
4259,6,96680767841,Ornare Limited,motor vehicle supplies and new parts,85.723766,99.850969,98.435171,63.772083,90.097039,48.907104
3772,7,86578477987,Leo In Consulting,"watch, clock, and jewelry repair shops",85.650047,99.925484,99.975161,55.113212,95.172929,53.800298
1956,8,49212265466,Auctor Company,"florists supplies, nursery stock, and flowers",85.552242,99.056135,98.782911,60.014929,94.575765,49.503229
2367,9,57699602880,Tellus Id Institute,"stationery, office supplies and printing and w...",85.490283,97.441629,82.103825,66.857427,97.885046,76.775956
1917,10,48534649627,Dignissim Maecenas Foundation,"opticians, optical goods, and eyeglasses",85.327189,99.975161,99.428713,59.119184,89.997512,51.539990


In [9]:
preliminary_top_20.to_csv(
    PROJECT_ROOT / "member5_ranking" / "results" / "preliminary_top_20.csv",
    index=False
)

Unavailable growth/stability scores are filled with the existing neutral percentile value 50.
This is an assumption for missing estimates, not a guarantee of unbiased treatment. The broader
low-sample flag does not itself alter finite scores, and OOD does not impose an automatic penalty.

## 7. Sensitivity Analysis

To test whether the merchant ranking is overly dependent on one specific weighting choice, three reasonable weighting scenarios are compared.

The scenarios represent:
- a balanced business view,
- a stronger emphasis on merchant value,
- a stronger emphasis on historical growth.

Ranking robustness is assessed using Top 100 overlap and merchant rank stability.

In [10]:
SCENARIOS = {
    "balanced": {
        "value_score": 0.30,
        "customer_score": 0.25,
        "growth_score": 0.20,
        "stability_score": 0.15,
        "market_score": 0.10,
    },
    "value_focused": {
        "value_score": 0.40,
        "customer_score": 0.25,
        "growth_score": 0.15,
        "stability_score": 0.15,
        "market_score": 0.05,
    },
    "growth_focused": {
        "value_score": 0.25,
        "customer_score": 0.20,
        "growth_score": 0.30,
        "stability_score": 0.15,
        "market_score": 0.10,
    },
}

for scenario_name, weights in SCENARIOS.items():
    score_col = f"{scenario_name}_score"
    rank_col = f"{scenario_name}_rank"

    score_df[score_col] = sum(
        weight * score_df[column]
        for column, weight in weights.items()
    )

    score_df[rank_col] = (
        score_df[score_col]
        .rank(method="min", ascending=False)
        .astype(int)
    )

In [11]:
top100_sets = {}

for scenario_name in SCENARIOS:
    score_col = f"{scenario_name}_score"

    top100_sets[scenario_name] = set(
        score_df
        .nlargest(100, score_col)["merchant_abn"]
    )

overlap_results = []

scenario_names = list(SCENARIOS.keys())

for i in range(len(scenario_names)):
    for j in range(i + 1, len(scenario_names)):

        scenario_a = scenario_names[i]
        scenario_b = scenario_names[j]

        overlap_count = len(
            top100_sets[scenario_a]
            & top100_sets[scenario_b]
        )

        overlap_results.append({
            "scenario_a": scenario_a,
            "scenario_b": scenario_b,
            "top100_overlap_count": overlap_count,
            "top100_overlap_rate": overlap_count / 100,
        })

overlap_df = pd.DataFrame(overlap_results)

overlap_df

,scenario_a,scenario_b,top100_overlap_count,top100_overlap_rate
0,balanced,value_focused,85,0.85
1,balanced,growth_focused,80,0.80
2,value_focused,growth_focused,66,0.66


In [12]:
rank_columns = [
    "balanced_rank",
    "value_focused_rank",
    "growth_focused_rank",
]

score_df["best_rank"] = score_df[rank_columns].min(axis=1)
score_df["worst_rank"] = score_df[rank_columns].max(axis=1)

score_df["rank_range"] = (
    score_df["worst_rank"]
    - score_df["best_rank"]
)

rank_stability = (
    score_df[
        [
            "merchant_abn",
            "merchant_name",
            "balanced_rank",
            "value_focused_rank",
            "growth_focused_rank",
            "rank_range",
        ]
    ]
    .sort_values("rank_range")
)

rank_stability.head(20)

,merchant_abn,merchant_name,balanced_rank,value_focused_rank,growth_focused_rank,rank_range
4236,96190048310,Penatibus Et Inc.,4023,4023,4023,0
178,13747603419,Fermentum Vel Mauris Institute,4024,4024,4024,0
3034,71475747855,Aliquet Molestie Corporation,4025,4025,4025,0
985,29623808496,Iaculis Odio Nam Foundation,4026,4026,4026,0
4144,93915598279,Molestie Pharetra Nibh LLP,4022,4021,4022,1
373,17507773571,Libero Nec Limited,4020,4019,4020,1
2745,65959377833,Lobortis Augue Ltd,4021,4022,4021,1
3608,82882460979,Sem Molestie Foundation,4011,4010,4011,1
280,15704713883,Faucibus Ut Nulla Ltd,4016,4016,4017,1
3994,91062920626,Ac Inc.,4019,4018,4019,1


In [13]:
top100_stability = (
    score_df[score_df["balanced_rank"] <= 100]
    [
        [
            "merchant_abn",
            "merchant_name",
            "balanced_rank",
            "value_focused_rank",
            "growth_focused_rank",
            "rank_range",
        ]
    ]
    .sort_values("balanced_rank")
)

top100_stability

,merchant_abn,merchant_name,balanced_rank,value_focused_rank,growth_focused_rank,rank_range
3966,90568944804,Diam Eu Dolor LLC,1,4,1,3
1414,38090089066,Interdum Feugiat Sed Inc.,2,1,4,3
869,27326652377,Tellus Aenean Corporation,3,13,2,11
3782,86772484982,Posuere Cubilia Curae LLC,4,28,8,24
3692,84703983173,Amet Consulting,5,8,5,3
...,...,...,...,...,...,...
989,29639699851,Sodales Elit Erat Corporation,96,87,135,48
2696,64732735902,Imperdiet Non Vestibulum Institute,97,138,82,56
3348,77590625261,Sed Diam Foundation,98,76,189,113
1300,35556933338,Semper Cursus Integer Limited,99,90,130,40


In [14]:
top100_stability["rank_range"].describe()

count    100.000000
mean      49.510000
std       32.408533
min        3.000000
25%       25.500000
50%       44.000000
75%       70.250000
max      179.000000
Name: rank_range, dtype: float64

In [15]:
top100_most_sensitive = (
    top100_stability
    .sort_values("rank_range", ascending=False)
    .head(20)
)

top100_most_sensitive

,merchant_abn,merchant_name,balanced_rank,value_focused_rank,growth_focused_rank,rank_range
2226,55069630872,A Nunc Corp.,93,200,21,179
4119,93558142492,Dolor Quisque Inc.,94,56,183,127
1969,49505931725,Suspendisse Ac Associates,71,42,164,122
3348,77590625261,Sed Diam Foundation,98,76,189,113
3085,72472909171,Nullam Consulting,73,43,154,111
4175,94729574738,Scelerisque Corporation,72,115,10,105
2396,58454491168,Diam At Foundation,62,40,140,100
2471,60111071436,Imperdiet Non LLC,81,141,45,96
3479,80324045558,Ipsum Dolor Sit Corporation,84,59,155,96
1128,32361057556,Orci In Consequat Corporation,58,31,125,94


In [16]:
top20_stability = (
    score_df[score_df["balanced_rank"] <= 20]
    [
        [
            "merchant_abn",
            "merchant_name",
            "balanced_rank",
            "value_focused_rank",
            "growth_focused_rank",
            "rank_range",
        ]
    ]
    .sort_values("balanced_rank")
)

top20_stability

,merchant_abn,merchant_name,balanced_rank,value_focused_rank,growth_focused_rank,rank_range
3966,90568944804,Diam Eu Dolor LLC,1,4,1,3
1414,38090089066,Interdum Feugiat Sed Inc.,2,1,4,3
869,27326652377,Tellus Aenean Corporation,3,13,2,11
3782,86772484982,Posuere Cubilia Curae LLC,4,28,8,24
3692,84703983173,Amet Consulting,5,8,5,3
4259,96680767841,Ornare Limited,6,3,9,6
3772,86578477987,Leo In Consulting,7,2,32,30
1956,49212265466,Auctor Company,8,5,14,9
2367,57699602880,Tellus Id Institute,9,52,3,49
1917,48534649627,Dignissim Maecenas Foundation,10,6,25,19


In [17]:
# Number of balanced Top 100 merchants that remain Top 100 in all scenarios
stable_top100_count = (
    (score_df["balanced_rank"] <= 100)
    & (score_df["value_focused_rank"] <= 100)
    & (score_df["growth_focused_rank"] <= 100)
).sum()

print("Balanced Top 100 retained in all scenarios:", stable_top100_count)

Balanced Top 100 retained in all scenarios: 66


In [18]:
# Summary for the balanced Top 20
top20_stability["rank_range"].describe()

count    20.000000
mean     18.250000
std      13.466821
min       3.000000
25%       6.750000
50%      17.000000
75%      26.000000
max      49.000000
Name: rank_range, dtype: float64

### Business-only sensitivity interpretation

These three scenarios precede fraud integration. Top100 overlap is 85 between balanced and
value-focused, 80 between balanced and growth-focused, and 66 between value-focused and growth-focused.
There are 66 merchants shared across all three **business-only** selections; this is not a verified
robust core of the final fraud-adjusted ranking. Section 12 separately applies these same business scenarios with the baseline fraud adjustment.

For the balanced business Top100, median rank range is 44 positions; for its Top20 it is 17.
These diagnostics describe dependence on commercial preferences on the same historical dataset.
They neither establish future performance nor prove that the baseline is the correct ranking.

## 8. Fraud Risk Integration

Member 4's `member4_fraud/result/knn_fraud_risk_scores.csv` already combines consumer fraud exposure and KNN-estimated merchant fraud risk.

The balanced baseline is:

`fraud_risk_index = 0.50 * consumer_exposure_percentile + 0.50 * knn_merchant_risk_percentile`

`risk_safety_score = 100 * (1 - fraud_risk_index)`

The 50/50 composition is a balanced baseline, not a statistically optimal weight. These are relative fraud-risk signals, not precise fraud probabilities. The formula applies where both components are available. Seven eligible merchants lack consumer exposure; Member 4 already supplies KNN-only combined scores for these rows. Member 5 consumes the supplied scores unchanged and performs no missing-fraud filling. Missing combined scores or reliability diagnostics stop execution.

In [19]:
FRAUD_PATH = (
    PROJECT_ROOT
    / "member4_fraud"
    / "result"
    / "knn_fraud_risk_scores.csv"
)

fraud_df = pd.read_csv(FRAUD_PATH, dtype={"merchant_abn": str})

print("Project root:", PROJECT_ROOT)
print("Fraud profile shape:", fraud_df.shape)
print("Unique merchants:", fraud_df["merchant_abn"].nunique())

fraud_df.head()

Project root: /Users/xiexinyu/Desktop/MAST30034-Project-2-Group-50
Fraud profile shape: (4422, 19)
Unique merchants: 4422


,merchant_abn,merchant_name,consumer_exposure_percentile,knn_score,knn_merchant_risk_percentile,knn_mean_neighbor_distance,knn_max_neighbor_distance,knn_distance_percentile,knn_confidence_score,knn_out_of_distribution,knn_score_source,fraud_risk_70c_30knn,risk_safety_70c_30knn,fraud_risk_50c_50knn,risk_safety_50c_50knn,fraud_risk_30c_70knn,risk_safety_30c_70knn,fraud_risk_index,risk_safety_score
0,10023283211,Felis Limited,0.579878,0.509525,0.334314,4.486615,5.220752,0.967213,0.032787,True,full_model_for_unlabelled_merchant,0.506208,49.379160,0.457096,54.290443,0.407983,59.201726,0.457096,54.290443
1,10142254217,Arcu Ac Orci Corporation,0.447315,0.657310,0.692830,3.447449,4.239421,0.967213,0.032787,True,full_model_for_unlabelled_merchant,0.520969,47.903077,0.570072,42.992779,0.619175,38.082480,0.570072,42.992779
2,10165489824,Nunc Sed Company,0.000000,0.313441,0.111061,3.130178,4.717325,0.934426,0.065574,False,full_model_for_unlabelled_merchant,0.033318,96.668175,0.055530,94.446958,0.077743,92.225741,0.055530,94.446958
3,10187291046,Ultricies Dignissim Lacus Foundation,0.114888,0.658932,0.765890,1.129005,1.470959,0.426230,0.573770,False,full_model_for_unlabelled_merchant,0.310189,68.981150,0.440389,55.961105,0.570589,42.941060,0.440389,55.961105
4,10192359162,Enim Condimentum PC,0.902334,0.652009,0.566614,1.489107,2.008204,0.622951,0.377049,False,full_model_for_unlabelled_merchant,0.801618,19.838202,0.734474,26.552605,0.667330,33.267007,0.734474,26.552605


In [20]:
fraud_columns = [
    "merchant_abn",
    "fraud_risk_index",
    "risk_safety_score",

    "fraud_risk_70c_30knn",
    "risk_safety_70c_30knn",

    "fraud_risk_50c_50knn",
    "risk_safety_50c_50knn",

    "fraud_risk_30c_70knn",
    "risk_safety_30c_70knn",

    "consumer_exposure_percentile",
    "knn_merchant_risk_percentile",
    "knn_confidence_score",
    "knn_out_of_distribution",
]

fraud_for_ranking = fraud_df[fraud_columns].copy()

required_fraud_columns = [column for column in fraud_columns if column != "consumer_exposure_percentile"]
if fraud_for_ranking[required_fraud_columns].isna().any().any():
    raise ValueError("Required KNN fraud inputs contain missing values.")
if fraud_for_ranking["merchant_abn"].duplicated().any():
    raise ValueError("KNN fraud inputs contain duplicate merchant identifiers.")
for suffix, consumer_weight in [("70c_30knn", 0.70), ("50c_50knn", 0.50), ("30c_70knn", 0.30)]:
    expected_risk = (
        consumer_weight * fraud_for_ranking["consumer_exposure_percentile"]
        + (1 - consumer_weight) * fraud_for_ranking["knn_merchant_risk_percentile"]
    )
    both_available = expected_risk.notna()
    np.testing.assert_allclose(
        fraud_for_ranking.loc[both_available, f"fraud_risk_{suffix}"],
        expected_risk.loc[both_available],
    )
    np.testing.assert_allclose(
        fraud_for_ranking[f"risk_safety_{suffix}"], 100 * (1 - fraud_for_ranking[f"fraud_risk_{suffix}"])
    )
np.testing.assert_allclose(fraud_for_ranking["fraud_risk_index"], fraud_for_ranking["fraud_risk_50c_50knn"])
np.testing.assert_allclose(fraud_for_ranking["risk_safety_score"], fraud_for_ranking["risk_safety_50c_50knn"])

fraud_for_ranking.to_csv(
    PROJECT_ROOT
    / "member5_ranking"
    / "results"
    / "fraud_for_ranking.csv",
    index=False
)

print("Shape:", fraud_for_ranking.shape)
fraud_for_ranking.head()

Shape: (4422, 13)


,merchant_abn,fraud_risk_index,risk_safety_score,fraud_risk_70c_30knn,risk_safety_70c_30knn,fraud_risk_50c_50knn,risk_safety_50c_50knn,fraud_risk_30c_70knn,risk_safety_30c_70knn,consumer_exposure_percentile,knn_merchant_risk_percentile,knn_confidence_score,knn_out_of_distribution
0,10023283211,0.457096,54.290443,0.506208,49.379160,0.457096,54.290443,0.407983,59.201726,0.579878,0.334314,0.032787,True
1,10142254217,0.570072,42.992779,0.520969,47.903077,0.570072,42.992779,0.619175,38.082480,0.447315,0.692830,0.032787,True
2,10165489824,0.055530,94.446958,0.033318,96.668175,0.055530,94.446958,0.077743,92.225741,0.000000,0.111061,0.065574,False
3,10187291046,0.440389,55.961105,0.310189,68.981150,0.440389,55.961105,0.570589,42.941060,0.114888,0.765890,0.573770,False
4,10192359162,0.734474,26.552605,0.801618,19.838202,0.734474,26.552605,0.667330,33.267007,0.902334,0.566614,0.377049,False


In [21]:
score_df["merchant_abn"] = score_df["merchant_abn"].astype(str)
fraud_for_ranking["merchant_abn"] = fraud_for_ranking["merchant_abn"].astype(str)

final_df = score_df.merge(
    fraud_for_ranking,
    on="merchant_abn",
    how="left",
    validate="one_to_one",
)

print("Before merge:", len(score_df))
print("After merge:", len(final_df))

print(
    "Missing fraud risk:",
    final_df["fraud_risk_index"].isna().sum()
)

if final_df[required_fraud_columns].isna().any().any():
    raise ValueError("Eligible merchants are missing required KNN fraud inputs.")
if final_df["balanced_score"].isna().any():
    raise ValueError("Eligible merchants are missing business scores.")

print("Eligible merchants without consumer exposure (upstream scores retained):",
      final_df["consumer_exposure_percentile"].isna().sum())

Before merge: 4026
After merge: 4026
Missing fraud risk: 0
Eligible merchants without consumer exposure (upstream scores retained): 7


In [22]:
final_df["risk_safety_score"].describe()

count    4026.000000
mean       49.143295
std        20.731964
min         1.382075
25%        34.189044
50%        48.940766
75%        65.137231
max        99.762374
Name: risk_safety_score, dtype: float64

## 9. Final Ranking with Fraud Risk

The final merchant ranking combines business performance with fraud-risk safety.

The business component uses the balanced business score developed earlier. Fraud risk is converted into a safety score, where higher values indicate lower relative fraud risk.

The baseline final score assigns 90% weight to business performance and 10% weight to fraud-risk safety.

In [23]:
final_df["final_score"] = (
    0.90 * final_df["balanced_score"]
    + 0.10 * final_df["risk_safety_score"]
)

final_df["final_rank"] = (
    final_df["final_score"]
    .rank(method="min", ascending=False)
    .astype(int)
)


final_df[
    [
        "merchant_abn",
        "merchant_name",
        "balanced_score",
        "risk_safety_score",
        "final_score",
        "balanced_rank",
        "final_rank",
    ]
].sort_values("final_rank").head()

,merchant_abn,merchant_name,balanced_score,risk_safety_score,final_score,balanced_rank,final_rank
1301,38090089066,Interdum Feugiat Sed Inc.,86.800764,69.637978,85.084485,2,1
3614,90568944804,Diam Eu Dolor LLC,87.168863,63.826405,84.834617,1,2
3369,84703983173,Amet Consulting,85.765124,74.419880,84.630599,5,3
3443,86578477987,Leo In Consulting,85.650047,74.543876,84.539430,7,4
803,27326652377,Tellus Aenean Corporation,86.253316,67.360254,84.364009,3,5


In [24]:
final_df["rank_change_after_fraud"] = (
    final_df["balanced_rank"] - final_df["final_rank"]
)

final_df["rank_change_after_fraud"].describe()

count    4026.000000
mean        0.000000
std       113.404551
min      -299.000000
25%       -71.750000
50%        -3.000000
75%        63.000000
max       415.000000
Name: rank_change_after_fraud, dtype: float64

### Fraud-composition sensitivity

Hold the final fraud-safety weight at 10% and compare 70% consumer / 30% KNN, 50% / 50%, and 30% / 70%. The 50/50 composition is a balanced baseline, not a statistically optimal mix.


In [25]:
final_df["final_score_70c_30knn"] = (
    0.90 * final_df["balanced_score"]
    + 0.10 * final_df["risk_safety_70c_30knn"]
)

final_df["final_score_50c_50knn"] = (
    0.90 * final_df["balanced_score"]
    + 0.10 * final_df["risk_safety_50c_50knn"]
)

final_df["final_score_30c_70knn"] = (
    0.90 * final_df["balanced_score"]
    + 0.10 * final_df["risk_safety_30c_70knn"]
)

In [26]:
np.testing.assert_allclose(final_df["final_score_50c_50knn"], final_df["final_score"])

In [27]:
top100_70 = set(
    final_df.nlargest(100, "final_score_70c_30knn")["merchant_abn"]
)

top100_50 = set(
    final_df.nlargest(100, "final_score_50c_50knn")["merchant_abn"]
)

top100_30 = set(
    final_df.nlargest(100, "final_score_30c_70knn")["merchant_abn"]
)


rank_70 = final_df["final_score_70c_30knn"].rank()
rank_50 = final_df["final_score_50c_50knn"].rank()
rank_30 = final_df["final_score_30c_70knn"].rank()


sensitivity_summary = pd.DataFrame({
    "comparison": [
        "70/30 vs 50/50",
        "30/70 vs 50/50",
        "70/30 vs 30/70",
    ],
    "top100_overlap": [
        len(top100_70 & top100_50),
        len(top100_30 & top100_50),
        len(top100_70 & top100_30),
    ],
    "spearman_rank_correlation": [
        rank_70.corr(rank_50),
        rank_30.corr(rank_50),
        rank_70.corr(rank_30),
    ],
})

sensitivity_summary

,comparison,top100_overlap,spearman_rank_correlation
0,70/30 vs 50/50,96,0.999373
1,30/70 vs 50/50,95,0.999370
2,70/30 vs 30/70,91,0.997521


In [28]:
sensitivity_summary.to_csv(
    PROJECT_ROOT
    / "member5_ranking"
    / "results"
    / "fraud_composition_final_ranking_sensitivity.csv",
    index=False
)

Top 100 overlaps are 96 (70/30 vs 50/50), 95 (30/70 vs 50/50), and 91 (70/30 vs 30/70). Corresponding Spearman correlations are approximately 0.99937, 0.99937 and 0.99752. Overall ordering is very similar, with some selection changes near the Top 100 boundary.

### KNN reliability diagnostics


In [29]:
top100_final = (
    final_df
    .sort_values("final_rank")
    .head(100)
    .copy()
)

print(
    "Top 100 KNN out-of-distribution count:",
    top100_final["knn_out_of_distribution"].sum()
)

print(
    "Top 100 KNN out-of-distribution share:",
    top100_final["knn_out_of_distribution"].mean()
)

print(
    "Top 100 mean KNN confidence:",
    top100_final["knn_confidence_score"].mean()
)

print(
    "All eligible mean KNN confidence:",
    final_df["knn_confidence_score"].mean()
)

Top 100 KNN out-of-distribution count: 12
Top 100 KNN out-of-distribution share: 0.12
Top 100 mean KNN confidence: 0.44196721311475407
All eligible mean KNN confidence: 0.19361852874349514


In [30]:
top100_final[
    [
        "merchant_abn",
        "merchant_name",
        "final_rank",
        "final_score",
        "knn_confidence_score",
        "knn_out_of_distribution",
    ]
].head(20)

,merchant_abn,merchant_name,final_rank,final_score,knn_confidence_score,knn_out_of_distribution
1301,38090089066,Interdum Feugiat Sed Inc.,1,85.084485,0.918033,False
3614,90568944804,Diam Eu Dolor LLC,2,84.834617,0.327869,False
3369,84703983173,Amet Consulting,3,84.630599,0.459016,False
3443,86578477987,Leo In Consulting,4,84.539430,0.639344,False
803,27326652377,Tellus Aenean Corporation,5,84.364009,0.245902,False
1401,40515428545,Elit Sed Consequat Associates,6,84.276448,0.557377,False
348,17488304283,Posuere Cubilia Curae Corporation,7,84.273831,0.377049,False
1798,49212265466,Auctor Company,8,84.246497,0.885246,False
3453,86772484982,Posuere Cubilia Curae LLC,9,84.237758,0.065574,False
3881,96680767841,Ornare Limited,10,84.216013,0.967213,False


### Fraud-model reliability in the final recommendations

The KNN fraud model also provides a confidence diagnostic based on merchant similarity to the labelled training sample.

Among the final Top 100 merchants:

- 12 merchants (12%) are flagged as out-of-distribution;
- the mean KNN confidence score is 0.442, compared with 0.194 across all eligible merchants;
- none of the Top 20 merchants are flagged as out-of-distribution.

This suggests that the highest-ranked recommendations generally have stronger support from the KNN similarity structure than the wider merchant population. However, fraud-risk estimates for the 12 out-of-distribution merchants should be interpreted with greater caution rather than treated as precise fraud probabilities.

In [31]:
reliability_summary = pd.DataFrame({
    "metric": [
        "top100_ood_count",
        "top100_ood_share",
        "top100_mean_knn_confidence",
        "all_eligible_mean_knn_confidence",
        "top20_ood_count",
    ],
    "value": [
        int(top100_final["knn_out_of_distribution"].sum()),
        top100_final["knn_out_of_distribution"].mean(),
        top100_final["knn_confidence_score"].mean(),
        final_df["knn_confidence_score"].mean(),
        int(top100_final.head(20)["knn_out_of_distribution"].sum()),
    ],
})

reliability_summary.to_csv(
    PROJECT_ROOT
    / "member5_ranking"
    / "results"
    / "top100_knn_reliability_summary.csv",
    index=False
)

reliability_summary

,metric,value
0,top100_ood_count,12.000000
1,top100_ood_share,0.120000
2,top100_mean_knn_confidence,0.441967
3,all_eligible_mean_knn_confidence,0.193619
4,top20_ood_count,0.000000


In [32]:
largest_rank_changes = (
    final_df[
        [
            "merchant_abn",
            "merchant_name",
            "balanced_rank",
            "final_rank",
            "rank_change_after_fraud",
            "balanced_score",
            "risk_safety_score",
            "knn_confidence_score",
            "knn_out_of_distribution",
        ]
    ]
    .assign(
        absolute_rank_change=lambda x:
        x["rank_change_after_fraud"].abs()
    )
    .sort_values("absolute_rank_change", ascending=False)
    .head(20)
)

largest_rank_changes

,merchant_abn,merchant_name,balanced_rank,final_rank,rank_change_after_fraud,balanced_score,risk_safety_score,knn_confidence_score,knn_out_of_distribution,absolute_rank_change
3371,84787662573,Varius Orci Inc.,2486,2071,415,42.920730,99.286794,0.032787,True,415
3123,79081516258,A Neque LLP,2633,2224,409,40.513710,99.570130,0.032787,True,409
1338,38918664617,Faucibus Foundation,2466,2062,404,43.238587,97.158508,0.032787,True,404
3097,78577106740,Donec Consulting,2464,2061,403,43.243667,97.213129,0.032787,True,403
102,12240378188,Ut Ipsum Corp.,2636,2236,400,40.473307,98.471720,0.032787,True,400
3549,88745863985,Fringilla Ornare Placerat Institute,2978,2588,390,35.101828,98.596495,0.032787,True,390
1931,52448445000,Feugiat Placerat Industries,2631,2246,385,40.518563,96.818214,0.032787,True,385
573,22503967537,Iaculis Quis LLC,3088,2711,377,33.074534,99.230472,0.032787,True,377
3570,89133730546,Adipiscing Fringilla Porttitor Ltd,3100,2725,375,32.900993,98.664230,0.032787,True,375
990,31416331470,Quis Turpis Corporation,3037,2663,374,33.906411,98.653372,0.032787,True,374


In [33]:
business_top100 = set(
    final_df.loc[final_df["balanced_rank"] <= 100, "merchant_abn"]
)

final_top100 = set(
    final_df.loc[final_df["final_rank"] <= 100, "merchant_abn"]
)

top100_overlap = len(business_top100 & final_top100)

print("Business vs final Top 100 overlap:", top100_overlap)
print("Overlap rate:", top100_overlap / 100)

Business vs final Top 100 overlap: 90
Overlap rate: 0.9


### Initial Fraud Adjustment Check

The overlap above measures how the 10% fraud-safety adjustment changes business-only Top 100 membership. The rank-change table retains KNN confidence and out-of-distribution diagnostics to help interpret relative risk signals.

Business performance retains 90% of the final score. Fraud safety is a baseline adjustment; the sensitivity analyses assess how selection changes under alternative risk assumptions.

## 10. Sensitivity to Fraud Weight

The baseline final ranking assigns 10% weight to fraud-risk safety.

To test whether the final merchant selection is overly dependent on this assumption, alternative fraud weights of 5% and 15% are compared with the 10% baseline.

In [34]:
FRAUD_WEIGHTS = {
    "fraud_5": 0.05,
    "fraud_10": 0.10,
    "fraud_15": 0.15,
}

for name, fraud_weight in FRAUD_WEIGHTS.items():
    business_weight = 1 - fraud_weight

    score_col = f"{name}_score"
    rank_col = f"{name}_rank"

    final_df[score_col] = (
        business_weight * final_df["balanced_score"]
        + fraud_weight * final_df["risk_safety_score"]
    )

    final_df[rank_col] = (
        final_df[score_col]
        .rank(method="min", ascending=False)
        .astype(int)
    )

In [35]:
fraud_top100_sets = {}

for name in FRAUD_WEIGHTS:
    rank_col = f"{name}_rank"

    fraud_top100_sets[name] = set(
        final_df.loc[
            final_df[rank_col] <= 100,
            "merchant_abn"
        ]
    )

fraud_overlap_results = []

scenario_names = list(FRAUD_WEIGHTS.keys())

for i in range(len(scenario_names)):
    for j in range(i + 1, len(scenario_names)):
        a = scenario_names[i]
        b = scenario_names[j]

        overlap = len(
            fraud_top100_sets[a]
            & fraud_top100_sets[b]
        )

        fraud_overlap_results.append({
            "scenario_a": a,
            "scenario_b": b,
            "top100_overlap_count": overlap,
            "top100_overlap_rate": overlap / 100,
        })

fraud_weight_overlap_df = pd.DataFrame(
    fraud_overlap_results
)

fraud_weight_overlap_df

,scenario_a,scenario_b,top100_overlap_count,top100_overlap_rate
0,fraud_5,fraud_10,94,0.94
1,fraud_5,fraud_15,86,0.86
2,fraud_10,fraud_15,92,0.92


In [36]:
fraud_rank_cols = [
    "fraud_5_rank",
    "fraud_10_rank",
    "fraud_15_rank",
]

final_df["fraud_weight_best_rank"] = (
    final_df[fraud_rank_cols].min(axis=1)
)

final_df["fraud_weight_worst_rank"] = (
    final_df[fraud_rank_cols].max(axis=1)
)

final_df["fraud_weight_rank_range"] = (
    final_df["fraud_weight_worst_rank"]
    - final_df["fraud_weight_best_rank"]
)

fraud_weight_top100_stability = (
    final_df[
        final_df["fraud_10_rank"] <= 100
    ][
        [
            "merchant_abn",
            "merchant_name",
            "fraud_5_rank",
            "fraud_10_rank",
            "fraud_15_rank",
            "fraud_weight_rank_range",
        ]
    ]
)

fraud_weight_top100_stability[
    "fraud_weight_rank_range"
].describe()

count    100.000000
mean      14.820000
std       13.982225
min        0.000000
25%        5.000000
50%       10.000000
75%       22.000000
max       62.000000
Name: fraud_weight_rank_range, dtype: float64

In [37]:
fraud_weight_most_sensitive = (
    fraud_weight_top100_stability
    .sort_values(
        "fraud_weight_rank_range",
        ascending=False
    )
    .head(20)
)

fraud_weight_most_sensitive

,merchant_abn,merchant_name,fraud_5_rank,fraud_10_rank,fraud_15_rank,fraud_weight_rank_range
2165,57699602880,Tellus Id Institute,27,53,89,62
1621,45380641195,Nisl Arcu Iaculis Incorporated,52,77,113,61
1657,46012371285,A Ultricies Inc.,62,86,116,54
9,10323485998,Nunc Inc.,106,85,58,48
500,21025433654,Lorem Foundation,112,89,64,48
2417,63817709749,A Felis Ullamcorper Industries,64,84,103,39
56,11237511112,Magna Institute,60,80,99,39
1432,41271931352,Ac Sem Ut Company,79,96,117,38
3386,85362139954,In Consequat Associates,67,88,104,37
3450,86710922099,Ac Urna Consulting,81,58,45,36


### Fraud-Weight Sensitivity Interpretation

Top 100 overlap is 94 between 5% and 10% fraud weights, 86 between 5% and 15%, and 92 between 10% and 15%.

Among baseline Top 100 merchants, the rank range across these scenarios has median 10, mean 14.82 and maximum 62. Membership is relatively stable, while some exact ranks remain sensitive to risk preferences.

The 10% fraud weight is retained as a baseline decision-support assumption, not a statistically optimal weight.

### Final weighting framework

The locked BusinessScore is `0.30 Value + 0.25 Customer Reach + 0.20 Growth + 0.15 Stability + 0.10 Market Context`.
FinalScore is `0.90 BusinessScore + 0.10 FraudSafety`, giving effective weights of
27%, 22.5%, 18%, 13.5%, 9%, and 10%, respectively. Larger component scores are preferred.

The baseline prioritises historical commercial value and reach, with historical growth and
stability as complementary preferences. Stability rewards lower observed volatility; it does
not impose a volatility veto. MarketScore uses IRSAD only and is a contextual assumption.
FraudSafety combines the supplied consumer-exposure and KNN-risk percentiles 50/50 when both
exist, preserving the upstream KNN-only fallback where consumer exposure is missing.

These are business judgments, not statistically optimal weights. FraudSafety is not a probability
of no fraud, and KNN confidence is not prediction accuracy. Existing sensitivity results quantify
preference dependence; they do not validate future profitability. No scoring weights, normalization,
eligibility policy, fraud composition, or ranking logic are changed by the recommendation enhancements.

In [38]:
final_df.to_csv(
    PROJECT_ROOT
    / "member5_ranking"
    / "results"
    / "ranking_analysis_base.csv",
    index=False
)

## Decision scope and stakeholder use

The weights encode a business preference, not a statistically optimal solution. FinalScore is a
decision-support ranking, not a profit prediction. A higher score means stronger alignment with
these chosen criteria within this eligible pool; it does not estimate the probability of successful onboarding.

| Assumption or limitation | How a stakeholder should use the ranking |
|---|---|
| `estimated_bnpl_revenue` is total observed transaction value × take rate / 100, across the supplied history (2021-02-28 to 2022-10-26). | Treat it as a historical commercial value proxy, not future revenue, incremental BNPL sales, or net profit. Costs, defaults and adoption are not modelled. |
| `unique_consumers` counts distinct historical customers per merchant. | Interpret CustomerScore as reach/scale, not repeat loyalty or new customers. Do not sum merchant counts to obtain a deduplicated portfolio customer total. |
| Growth is a normalized monthly regression slope; stability is monthly revenue CV, using full months 2021-03 through 2022-09 and zero-filled gaps within each merchant's first-to-last activity span. | A slope is not a month-on-month percentage or forecast. The activity span does not include later silent months after the merchant's last transaction. Historical stability is not a guarantee of continued trading. |
| Fraud behavioural features/exposure use transactions through 2022-02-28, earlier than the business feature history. Census/SEIFA refer to 2021; ATO refers to 2021-22 and was published later. | This is a retrospective, mixed-window shortlist, not a common-date forward backtest. Refresh risk and business evidence before a live onboarding decision. |
| MarketScore positively ranks the transaction-weighted IRSAD national decile of customer postcodes. Higher regional advantage is a team-chosen contextual preference, not a proven driver of BNPL success. | Customer postcode context is neither merchant location nor an individual's actual income or creditworthiness. Census and ATO are contextual diagnostics, not additional direct scoring dimensions. Do not infer that lower-income customers or areas are fraudulent. |
| External postcode matches are incomplete and coverage differs by state. A matched postcode does not guarantee all fields are usable. | Read count-weighted source coverage with regional values; similar overall coverage does not prove missingness is unbiased. |
| Fraud labels are supplied risk scores, not confirmed transaction-level fraud outcomes; only 61 merchants have direct merchant labels. | FraudSafety (`risk_safety_score`) is a relative safety index, not a probability of no fraud. Do not treat a model estimate as evidence that a merchant committed fraud. |
| `knn_confidence_score` is 1 minus a relative neighbour-distance percentile. OOD means mean neighbour distance exceeds the labelled sample's 95th-percentile reference threshold. | KNN confidence measures relative similarity/support, not prediction accuracy or a calibrated probability. Prioritize evidence checks for OOD merchants; non-OOD is not a safety certificate. |
| OOD and low-sample flags are diagnostics, not automatic penalties. Only unavailable growth/CV scores receive the existing neutral value 50; finite but low-sample estimates remain scored. | Review thin history or weak model support rather than automatically rejecting the merchant. The current Top100 contains no low-sample growth flags under the team's thresholds, not a proof that every estimate is precise. |
| 396 transaction-only merchants lack necessary merchant master data, including category/take rate. | Request missing information before comparable scoring. Exclusion for incomplete data does not establish that a merchant is unsuitable for onboarding. |
| Sensitivity measures changes under selected preferences on the same data. The earlier commercial scenarios are business-only; fraud-composition and fraud-weight scenarios concern final scores. | Robust membership is not proof that the ranking is the correct answer. The 66 business-only common merchants differ from the final-scenario intersection computed in section 12; neither establishes correctness. |

**Usage:** use commercially strong selected merchants with adequate observed history and stronger
model support as priority candidates for normal due diligence. Selected OOD merchants, merchants
with thin history, and cases whose membership changes in the existing sensitivity outputs merit
manual review of current activity and risk evidence. These are review cues, not new scores,
approval thresholds or automatic rejection rules. A merchant outside the 100-slot shortlist may
still be commercially attractive; a small score difference at the cutoff is not a meaningful
probability difference. Review can change a stakeholder's eventual decision, but this notebook
does not alter the locked ranking or allocate extra slots.

## 11. Useful diagnostic: why not select by historical revenue alone?

Official recommendations remain the overall Top100 and five segment Top10 lists. All analyses below
operate on copies and do not overwrite the official scores or lists. Revenue means historical
`estimated_bnpl_revenue`, not future revenue or net profit. Customer counts are per merchant;
we do not sum them into independent portfolio customers.

The value-only comparator takes the 100 highest revenue proxies from the same eligible pool.
Report medians and Q1/Q3 for both complete lists and their non-overlapping members. These are
in-sample preference trade-offs, not an out-of-sample test of which strategy performs best.

In [39]:
from IPython.display import display, Markdown
from itertools import combinations

useful = final_df.copy(deep=True)
locked_snapshot = final_df.copy(deep=True)
useful_metrics = ["estimated_bnpl_revenue", "unique_consumers", "monthly_revenue_cv",
                  "normalized_monthly_revenue_trend", "fraud_risk_index", "risk_safety_score"]
useful_components = ["value_score", "customer_score", "growth_score", "stability_score", "market_score", "risk_safety_score"]

def diagnostic_top100(frame, score_column):
    ordered = frame.sort_values([score_column, "merchant_abn"], ascending=[False, True])
    # A cutoff tie would require an explicitly reviewed tie policy; do not silently change membership.
    assert ordered.iloc[99][score_column] > ordered.iloc[100][score_column], "Review cutoff tie"
    return set(ordered.head(100)["merchant_abn"])

def distribution_rows(groups, metrics):
    rows = []
    for group_name, frame in groups.items():
        for metric in metrics:
            values = frame[metric].dropna()
            rows.append(dict(group=group_name, metric=metric, merchant_count=len(frame),
                             nonmissing_count=len(values), missing_count=int(frame[metric].isna().sum()),
                             p25=values.quantile(.25), median=values.median(), p75=values.quantile(.75)))
    return pd.DataFrame(rows)

baseline_ids = diagnostic_top100(useful, "final_score")
value_ids = diagnostic_top100(useful, "estimated_bnpl_revenue")
useful["value_only_rank"] = useful.estimated_bnpl_revenue.rank(method="min", ascending=False).astype(int)
useful["current_selected"] = useful.merchant_abn.isin(baseline_ids)
useful["value_only_selected"] = useful.merchant_abn.isin(value_ids)
value_overlap = len(baseline_ids & value_ids)
value_groups = {
    "Current final Top100": useful.loc[useful.current_selected],
    "Value-only Top100": useful.loc[useful.value_only_selected],
    "Current-final-only": useful.loc[useful.current_selected & ~useful.value_only_selected],
    "Value-only-only": useful.loc[useful.value_only_selected & ~useful.current_selected],
}
value_comparison = distribution_rows(value_groups, useful_metrics)
value_comparison["top100_overlap_count"] = value_overlap
value_comparison["group_historical_revenue_proxy_total"] = value_comparison["group"].map({name: frame.estimated_bnpl_revenue.sum() for name, frame in value_groups.items()})
value_comparison.to_csv(RESULTS_DIR / "value_only_baseline_comparison.csv", index=False)
value_membership = useful.loc[useful.current_selected | useful.value_only_selected,
    ["merchant_abn", "merchant_name", "current_selected", "value_only_selected", "final_rank", "value_only_rank"] + useful_metrics].copy()
value_membership.to_csv(RESULTS_DIR / "value_only_baseline_membership.csv", index=False)
display(Markdown(f"**Overlap: {value_overlap}/100; {100-value_overlap} merchants unique to each list.**"))
display(value_comparison.pivot(index="metric", columns="group", values="median").round(5))
display(value_comparison[["group", "metric", "p25", "median", "p75", "missing_count"]])

**Overlap: 58/100; 42 merchants unique to each list.**

group,Current final Top100,Current-final-only,Value-only Top100,Value-only-only
metric,,,,
estimated_bnpl_revenue,214181.91447,86262.07406,299534.83747,248993.76484
fraud_risk_index,0.28575,0.30020,0.28300,0.28052
monthly_revenue_cv,0.18699,0.18803,0.18800,0.19329
normalized_monthly_revenue_trend,0.01574,0.01646,0.01493,0.01370
risk_safety_score,71.42465,69.98040,71.70047,71.94816
unique_consumers,15737.00000,11542.50000,14136.50000,7668.50000


,group,metric,p25,median,p75,missing_count
0,Current final Top100,estimated_bnpl_revenue,89805.197635,214181.914473,396969.976427,0
1,Current final Top100,unique_consumers,9413.250000,15737.000000,20684.500000,0
2,Current final Top100,monthly_revenue_cv,0.182900,0.186991,0.193381,0
3,Current final Top100,normalized_monthly_revenue_trend,0.014955,0.015739,0.016671,0
4,Current final Top100,fraud_risk_index,0.256993,0.285753,0.315931,0
5,Current final Top100,risk_safety_score,68.406899,71.424651,74.300682,0
6,Value-only Top100,estimated_bnpl_revenue,227261.201693,299534.837468,408442.581581,0
7,Value-only Top100,unique_consumers,6709.250000,14136.500000,19893.000000,0
8,Value-only Top100,monthly_revenue_cv,0.183331,0.187998,0.195467,0
9,Value-only Top100,normalized_monthly_revenue_trend,0.013894,0.014930,0.015947,0


In [40]:
# Four contrasting, reproducible illustrative cases: two broad trade-offs and two boundary cases.
case_specs = [
    ("84703983173", "Current-final-only", "Highest final-ranked entrant", "Scale/growth trade-off"),
    ("79827781481", "Value-only-only", "Highest historical revenue merchant omitted by final rule", "Scale/growth trade-off"),
    ("83690644458", "Current-final-only", "Last selected final-only merchant", "Capacity boundary"),
    ("23338656015", "Value-only-only", "Closest excluded value-only merchant by final rank", "Capacity boundary"),
]
replacement_specs = pd.DataFrame(case_specs, columns=["merchant_abn", "membership", "selection_reason", "comparison_pair"])
replacement_cases = replacement_specs.merge(useful, on="merchant_abn", validate="one_to_one")
assert len(replacement_cases) == 4
assert replacement_cases.current_selected.ne(replacement_cases.value_only_selected).all()
effective = {"value_score": .27, "customer_score": .225, "growth_score": .18,
             "stability_score": .135, "market_score": .09, "risk_safety_score": .10}
for col, weight in effective.items():
    replacement_cases[col + "_points"] = replacement_cases[col] * weight
point_cols = [c + "_points" for c in effective]
np.testing.assert_allclose(replacement_cases[point_cols].sum(axis=1), replacement_cases.final_score)
replacement_cases["interpretation"] = [
    "Lower revenue proxy than the value-only cutoff, but broad reach and stronger historical growth help its final score; this is not evidence of future profit.",
    "Highest revenue proxy and low CV do not offset weaker growth and reach percentiles under current preferences; not excluded because fraud was proven.",
    "The combined score just clears the capacity boundary; not superior on every raw metric to excluded merchants.",
    "A high-revenue, broad-reach merchant narrowly misses the final shortlist; weighted trade-offs, not commercial unsuitability, explain exclusion.",
]
replacement_cases = replacement_cases[["comparison_pair", "membership", "selection_reason", "merchant_abn", "merchant_name",
    "value_only_rank", "final_rank", "final_score"] + useful_metrics + useful_components[:-1] + point_cols +
    ["knn_confidence_score", "knn_out_of_distribution", "interpretation"]]
replacement_cases.to_csv(RESULTS_DIR / "value_only_replacement_cases.csv", index=False)
display(replacement_cases[["merchant_name", "membership", "value_only_rank", "final_rank"] + useful_metrics])
for pair_name, pair in replacement_cases.groupby("comparison_pair", sort=False):
    entrant = pair.loc[pair.membership.eq("Current-final-only")].iloc[0]
    outgoing = pair.loc[pair.membership.eq("Value-only-only")].iloc[0]
    deltas = entrant[point_cols] - outgoing[point_cols]
    display(Markdown(f"**{pair_name}: {entrant.merchant_name} vs {outgoing.merchant_name}.** "
        f"Final-score gap {entrant.final_score-outgoing.final_score:+.4f}; weighted component-point differences "
        + ", ".join(f"{c.removesuffix('_score_points')}: {deltas[c]:+.3f}" for c in point_cols)
        + ". These pairs illustrate trade-offs; they are not literal one-to-one slot substitutions or causal attributions."))
med = value_comparison.pivot(index="metric", columns="group", values="median")
display(Markdown(f"""
**Decision value and cost:** the comprehensive rule replaces {100-value_overlap} revenue-only choices,
adding explicit reach, growth, volatility, contextual and relative-risk preferences. Historical revenue
proxy median is AUD {med.loc['estimated_bnpl_revenue','Current final Top100']:,.0f} versus
AUD {med.loc['estimated_bnpl_revenue','Value-only Top100']:,.0f}; reach is
{med.loc['unique_consumers','Current final Top100']:,.1f} versus {med.loc['unique_consumers','Value-only Top100']:,.1f}.
CV is {med.loc['monthly_revenue_cv','Current final Top100']:.5f} versus {med.loc['monthly_revenue_cv','Value-only Top100']:.5f};
normalized trend is {med.loc['normalized_monthly_revenue_trend','Current final Top100']:.5f} versus
{med.loc['normalized_monthly_revenue_trend','Value-only Top100']:.5f}.

The current selection does **not** dominate: its median relative risk index
({med.loc['fraud_risk_index','Current final Top100']:.5f}) is slightly higher than revenue-only
({med.loc['fraud_risk_index','Value-only Top100']:.5f}), and the volatility difference is small.
Among replacement groups the same trade-off must be read from the separate rows, not inferred from
whole-list medians. The comprehensive score makes priorities auditable; it does not demonstrate
higher future profit or a universally better shortlist. Revenue-only maximizes this historical
additive proxy for 100 merchants; choosing other criteria has a real historical-value opportunity cost.
"""))

,merchant_name,membership,value_only_rank,final_rank,estimated_bnpl_revenue,unique_consumers,monthly_revenue_cv,normalized_monthly_revenue_trend,fraud_risk_index,risk_safety_score
0,Amet Consulting,Current-final-only,139,3,98171.548195,19839,0.189186,0.017986,0.255801,74.419880
1,Amet Risus Inc.,Value-only-only,1,238,663870.290228,4314,0.168664,0.009993,0.284701,71.529876
2,Id Erat Etiam Consulting,Current-final-only,134,100,100271.548508,18665,0.184563,0.014403,0.291249,70.875077
3,Iaculis LLC,Value-only-only,92,103,186766.224503,19470,0.192573,0.014616,0.282311,71.768889


**Scale/growth trade-off: Amet Consulting vs Amet Risus Inc..** Final-score gap +6.2931; weighted component-point differences value: -0.925, customer: +2.901, growth: +6.794, stability: -1.216, market: -1.549, risk_safety: +0.289. These pairs illustrate trade-offs; they are not literal one-to-one slot substitutions or causal attributions.

**Capacity boundary: Id Erat Etiam Consulting vs Iaculis LLC.** Final-score gap +0.0277; weighted component-point differences value: -0.282, customer: -0.045, growth: -0.228, stability: +0.685, market: -0.013, risk_safety: -0.089. These pairs illustrate trade-offs; they are not literal one-to-one slot substitutions or causal attributions.


**Decision value and cost:** the comprehensive rule replaces 42 revenue-only choices,
adding explicit reach, growth, volatility, contextual and relative-risk preferences. Historical revenue
proxy median is AUD 214,182 versus
AUD 299,535; reach is
15,737.0 versus 14,136.5.
CV is 0.18699 versus 0.18800;
normalized trend is 0.01574 versus
0.01493.

The current selection does **not** dominate: its median relative risk index
(0.28575) is slightly higher than revenue-only
(0.28300), and the volatility difference is small.
Among replacement groups the same trade-off must be read from the separate rows, not inferred from
whole-list medians. The comprehensive score makes priorities auditable; it does not demonstrate
higher future profit or a universally better shortlist. Revenue-only maximizes this historical
additive proxy for 100 merchants; choosing other criteria has a real historical-value opportunity cost.


## 12. Final-ranking robustness to existing business preferences

Reuse `SCENARIOS` exactly. Each final scenario is 90% of its business score plus 10% of the same
50/50 baseline FraudSafety, including the existing upstream missing-exposure fallback. Percentiles,
eligibility and risk signals are fixed. This isolates business preferences; it is not a joint sweep
of every risk/business assumption and is not evidence that any scenario is more correct.

Consistent means selected in all three scenarios. Sensitive means selected in one or two.
A near-cutoff flag denotes baseline ranks 80–120 (inclusive), a transparent review window, not a
new selection rule. The review list includes every sensitive candidate and baseline-selected
OOD/low-sample candidates; OOD plus sensitivity is highlighted without penalizing scores.

In [41]:
scenario_rank_cols, scenario_selected_cols = [], []
final_scenario_sets = {}
for scenario_name, weights in SCENARIOS.items():
    assert np.isclose(sum(weights.values()), 1)
    score_col = f"diagnostic_{scenario_name}_final_score"
    rank_col = f"diagnostic_{scenario_name}_final_rank"
    selected_col = f"diagnostic_{scenario_name}_selected"
    useful[score_col] = .9 * sum(useful[c] * w for c, w in weights.items()) + .1 * useful.risk_safety_score
    useful[rank_col] = useful[score_col].rank(method="min", ascending=False).astype(int)
    final_scenario_sets[scenario_name] = diagnostic_top100(useful, score_col)
    useful[selected_col] = useful.merchant_abn.isin(final_scenario_sets[scenario_name])
    scenario_rank_cols.append(rank_col)
    scenario_selected_cols.append(selected_col)
np.testing.assert_allclose(useful.diagnostic_balanced_final_score, useful.final_score)
assert final_scenario_sets['balanced'] == baseline_ids
useful["scenario_selection_count"] = useful[scenario_selected_cols].sum(axis=1)
useful["consistently_selected"] = useful.scenario_selection_count.eq(len(SCENARIOS))
useful["selection_sensitive"] = useful.scenario_selection_count.between(1, len(SCENARIOS)-1)
useful["final_scenario_best_rank"] = useful[scenario_rank_cols].min(axis=1)
useful["final_scenario_worst_rank"] = useful[scenario_rank_cols].max(axis=1)
useful["final_scenario_rank_range"] = useful.final_scenario_worst_rank - useful.final_scenario_best_rank
useful["near_baseline_cutoff"] = useful.final_rank.between(80,120)
useful["ood_and_selection_sensitive"] = useful.knn_out_of_distribution & useful.selection_sensitive
scenario_summary_rows = []
for a,b in combinations(SCENARIOS,2):
    scenario_summary_rows.append(dict(scenario_a=a, scenario_b=b,
        top100_overlap=len(final_scenario_sets[a]&final_scenario_sets[b]),
        spearman_rank_correlation=useful[f'diagnostic_{a}_final_score'].rank().corr(useful[f'diagnostic_{b}_final_score'].rank()),
        consistently_selected_count=int(useful.consistently_selected.sum()),
        selected_union_count=int(useful.scenario_selection_count.gt(0).sum()),
        selection_sensitive_count=int(useful.selection_sensitive.sum()),
        baseline_selected_sensitive_count=int((useful.current_selected & useful.selection_sensitive).sum()),
        near_cutoff_sensitive_count=int((useful.near_baseline_cutoff & useful.selection_sensitive).sum()),
        ood_and_sensitive_count=int(useful.ood_and_selection_sensitive.sum()),
        baseline_selected_ood_and_sensitive_count=int((useful.current_selected & useful.ood_and_selection_sensitive).sum()),
        baseline_top100_rank_range_median=useful.loc[useful.current_selected,'final_scenario_rank_range'].median(),
        baseline_top100_rank_range_max=useful.loc[useful.current_selected,'final_scenario_rank_range'].max()))
final_business_summary = pd.DataFrame(scenario_summary_rows)
final_business_summary.to_csv(RESULTS_DIR/'final_business_sensitivity_summary.csv',index=False)
scenario_detail_cols = ['merchant_abn','merchant_name','final_rank','current_selected'] + scenario_rank_cols + scenario_selected_cols + [
    'scenario_selection_count','consistently_selected','selection_sensitive','near_baseline_cutoff',
    'final_scenario_best_rank','final_scenario_worst_rank','final_scenario_rank_range',
    'knn_confidence_score','knn_out_of_distribution','low_sample_growth_estimate','ood_and_selection_sensitive']
scenario_details = useful[scenario_detail_cols].sort_values('final_rank')
scenario_details.to_csv(RESULTS_DIR/'final_business_sensitivity_merchants.csv',index=False)
review_mask = useful.selection_sensitive | (useful.current_selected & (useful.knn_out_of_distribution | useful.low_sample_growth_estimate))
review_list = useful.loc[review_mask,scenario_detail_cols].copy()
review_list['review_reason'] = review_list.apply(lambda x: '; '.join(
    label for flag,label in [(x.selection_sensitive,'Membership depends on business preferences'),
                             (x.knn_out_of_distribution,'OOD: weak relative model support'),
                             (x.low_sample_growth_estimate,'Thin growth history')]
    if flag),axis=1)
review_list['stakeholder_action'] = 'Review business priorities and current evidence; no automatic rejection, score penalty or extra slots.'
review_list = review_list.sort_values(['ood_and_selection_sensitive','current_selected','final_rank'],ascending=[False,False,True])
review_list.to_csv(RESULTS_DIR/'final_business_sensitivity_review_list.csv',index=False)
display(final_business_summary)
display(scenario_details.loc[scenario_details.consistently_selected,['merchant_name','final_rank']+scenario_rank_cols])
display(review_list)
display(Markdown(f"**{int(useful.consistently_selected.sum())} consistently selected**, "
    f"{int(useful.selection_sensitive.sum())} selected in only some scenarios across their union; "
    f"{int(useful.ood_and_selection_sensitive.sum())} combine OOD with sensitivity. "
    f"The review list contains {len(review_list)} candidates, including potential entrants; it is not an additional recommendation list. "
    "Robustness is conditional on these three scenarios, not correctness or future performance."))

,scenario_a,scenario_b,top100_overlap,spearman_rank_correlation,consistently_selected_count,selected_union_count,selection_sensitive_count,baseline_selected_sensitive_count,near_cutoff_sensitive_count,ood_and_sensitive_count,baseline_selected_ood_and_sensitive_count,baseline_top100_rank_range_median,baseline_top100_rank_range_max
0,balanced,value_focused,88,0.992157,63,137,74,37,33,29,8,39.0,121
1,balanced,growth_focused,75,0.986097,63,137,74,37,33,29,8,39.0,121
2,value_focused,growth_focused,63,0.963890,63,137,74,37,33,29,8,39.0,121


,merchant_name,final_rank,diagnostic_balanced_final_rank,diagnostic_value_focused_final_rank,diagnostic_growth_focused_final_rank
1301,Interdum Feugiat Sed Inc.,1,1,2,4
3614,Diam Eu Dolor LLC,2,2,21,1
3369,Amet Consulting,3,3,6,3
3443,Leo In Consulting,4,4,1,15
803,Tellus Aenean Corporation,5,5,29,2
...,...,...,...,...,...
2700,Mus Proin LLP,79,79,72,62
3147,Consectetuer Mauris Corp.,82,82,89,83
2417,A Felis Ullamcorper Industries,84,84,82,53
1604,Vel Pede Blandit Incorporated,87,87,100,41


,merchant_abn,merchant_name,final_rank,current_selected,diagnostic_balanced_final_rank,diagnostic_value_focused_final_rank,diagnostic_growth_focused_final_rank,diagnostic_balanced_selected,diagnostic_value_focused_selected,diagnostic_growth_focused_selected,...,near_baseline_cutoff,final_scenario_best_rank,final_scenario_worst_rank,final_scenario_rank_range,knn_confidence_score,knn_out_of_distribution,low_sample_growth_estimate,ood_and_selection_sensitive,review_reason,stakeholder_action
2165,57699602880,Tellus Id Institute,53,True,53,110,19,True,False,True,...,False,19,110,91,0.032787,True,False,True,Membership depends on business preferences; OO...,Review business priorities and current evidenc...
2461,64732735902,Imperdiet Non Vestibulum Institute,76,True,76,112,59,True,False,True,...,False,59,112,53,0.032787,True,False,True,Membership depends on business preferences; OO...,Review business priorities and current evidenc...
1621,45380641195,Nisl Arcu Iaculis Incorporated,77,True,77,131,68,True,False,True,...,False,68,131,63,0.032787,True,False,True,Membership depends on business preferences; OO...,Review business priorities and current evidenc...
3961,98416475066,Lacinia Orci PC,81,True,81,120,48,True,False,True,...,True,48,120,72,0.032787,True,False,True,Membership depends on business preferences; OO...,Review business priorities and current evidenc...
9,10323485998,Nunc Inc.,85,True,85,86,125,True,True,False,...,True,85,125,40,0.032787,True,False,True,Membership depends on business preferences; OO...,Review business priorities and current evidenc...
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
551,22033359776,Suspendisse Non Leo PC,130,False,130,76,194,False,True,False,...,False,76,194,118,0.950820,False,False,False,Membership depends on business preferences,Review business priorities and current evidenc...
69,11566786699,Euismod Et Institute,134,False,134,211,90,False,False,True,...,False,90,211,121,0.213115,False,False,False,Membership depends on business preferences,Review business priorities and current evidenc...
681,24938000144,Consequat Nec Corporation,137,False,137,171,98,False,False,True,...,False,98,171,73,0.098361,False,False,False,Membership depends on business preferences,Review business priorities and current evidenc...
2582,67400260923,Eleifend PC,153,False,153,94,254,False,True,False,...,False,94,254,160,0.639344,False,False,False,Membership depends on business preferences,Review business priorities and current evidenc...


**63 consistently selected**, 74 selected in only some scenarios across their union; 29 combine OOD with sensitivity. The review list contains 78 candidates, including potential entrants; it is not an additional recommendation list. Robustness is conditional on these three scenarios, not correctness or future performance.

## 13. Historical revenue concentration within the official Top100

Sort the already selected 100 merchants by historical revenue proxy solely to calculate cumulative
shares. Top5/10/20 below are **share statistics, not separate recommendation lists**. Segment shares
partition the same Top100. Concentration describes exposure to historical commercial scale;
there are no predicted profit, default-correlation or optimal-diversification claims.

In [42]:
selected_revenue = useful.loc[useful.current_selected].sort_values('estimated_bnpl_revenue',ascending=False).copy()
assert selected_revenue.estimated_bnpl_revenue.ge(0).all()
revenue_total = selected_revenue.estimated_bnpl_revenue.sum()
assert revenue_total > 0
concentration = pd.DataFrame([dict(largest_revenue_merchants=k, selected_pool_count=100,
    historical_revenue_proxy_sum=selected_revenue.head(k).estimated_bnpl_revenue.sum(),
    top100_historical_revenue_proxy_total=revenue_total,
    share_of_top100_historical_revenue_proxy=selected_revenue.head(k).estimated_bnpl_revenue.sum()/revenue_total)
    for k in [5,10,20]])
concentration.to_csv(RESULTS_DIR/'top100_revenue_concentration.csv',index=False)
industry_mapping = pd.read_csv(PROJECT_ROOT/'member3_industry_growth/results/category_to_group_mapping.csv')
print('Mapping columns:',industry_mapping.columns.tolist())
selected_industry = selected_revenue.merge(industry_mapping,on='merchant_category',how='left',validate='many_to_one')
assert selected_industry.industry_group.notna().all()
segment_revenue = selected_industry.groupby('industry_group').agg(
    top100_merchant_count=('merchant_abn','size'),
    historical_revenue_proxy_contribution=('estimated_bnpl_revenue','sum'),
    median_historical_revenue_proxy=('estimated_bnpl_revenue','median'))
segment_revenue['share_of_top100_revenue_proxy'] = segment_revenue.historical_revenue_proxy_contribution/revenue_total
assert len(segment_revenue)==5 and segment_revenue.top100_merchant_count.sum()==100
np.testing.assert_allclose(segment_revenue.share_of_top100_revenue_proxy.sum(),1)
segment_revenue.to_csv(RESULTS_DIR/'top100_segment_revenue_contribution.csv')
display(concentration)
display(segment_revenue.sort_values('share_of_top100_revenue_proxy',ascending=False))
largest_segment = segment_revenue.share_of_top100_revenue_proxy.idxmax()
display(Markdown("**Exposure:** "+', '.join(f"largest {int(x.largest_revenue_merchants)} account for {x.share_of_top100_historical_revenue_proxy:.1%}" for x in concentration.itertuples())
    +f" of the selected historical proxy. The largest segment is {largest_segment} at {segment_revenue.loc[largest_segment,'share_of_top100_revenue_proxy']:.1%}. "
    "The largest 20 merchants represent 20% of merchants but 40.7% of historical proxy: exposure is unequal, without the largest five dominating it. No single segment accounts for a majority in this snapshot. Selecting 100 merchants does not make exposure equal. "
    "These descriptive shares do not establish an acceptable risk limit. The ranking scores merchants independently and does not optimize portfolio diversification or correlated shocks; "
    "review concentration when planning onboarding, without interpreting these shares as future profit dependence."))

Mapping columns: ['merchant_category', 'industry_group']


,largest_revenue_merchants,selected_pool_count,historical_revenue_proxy_sum,top100_historical_revenue_proxy_total,share_of_top100_historical_revenue_proxy
0,5,100,3.027336e+06,2.549483e+07,0.118743
1,10,100,5.807752e+06,2.549483e+07,0.227801
2,20,100,1.036593e+07,2.549483e+07,0.406590


,top100_merchant_count,historical_revenue_proxy_contribution,median_historical_revenue_proxy,share_of_top100_revenue_proxy
industry_group,,,,
"Home, Garden & Living",25,6.434082e+06,255744.117365,0.252368
"Digital, Technology & Communications",30,6.258057e+06,97640.038539,0.245464
"Creative, Books & Leisure",13,4.368207e+06,305793.816710,0.171337
"Art, Gifts, Jewellery & Fashion",16,4.242210e+06,245595.898569,0.166395
"Mobility, Health & Specialist Services",16,4.192274e+06,204303.337537,0.164436


**Exposure:** largest 5 account for 11.9%, largest 10 account for 22.8%, largest 20 account for 40.7% of the selected historical proxy. The largest segment is Home, Garden & Living at 25.2%. The largest 20 merchants represent 20% of merchants but 40.7% of historical proxy: exposure is unequal, without the largest five dominating it. No single segment accounts for a majority in this snapshot. Selecting 100 merchants does not make exposure equal. These descriptive shares do not establish an acceptable risk limit. The ranking scores merchants independently and does not optimize portfolio diversification or correlated shocks; review concentration when planning onboarding, without interpreting these shares as future profit dependence.

## 14. MarketScore ablation (diagnostic only)

Baseline MarketScore weight remains 10% of BusinessScore (9% of FinalScore). For No-Market,
set its business weight to zero and divide the other four weights by 0.90. Keep the final
90/10 business/fraud mix, percentiles and all other inputs unchanged. No reduced-market scenario
is needed to answer this question.

Report all eligibility-pool Spearman rank correlation, switchers, and affected boundary candidates
(baseline or No-Market ranks 80–120, plus all switchers). Show raw transaction-weighted customer-area
IRSAD, its percentile score, and No-Market minus baseline score/rank changes. Percentile gaps describe
relative position, not proportional differences in raw regional context. An illustrative nearest-IRSAD
pair of opposite switchers helps assess numerical amplification without implying personal income.

In [43]:
no_market_weights = {c:w/(1-SCENARIOS['balanced']['market_score']) for c,w in SCENARIOS['balanced'].items() if c!='market_score'}
assert np.isclose(sum(no_market_weights.values()),1)
useful['no_market_final_score'] = .9*sum(useful[c]*w for c,w in no_market_weights.items())+.1*useful.risk_safety_score
useful['no_market_final_rank'] = useful.no_market_final_score.rank(method='min',ascending=False).astype(int)
no_market_ids = diagnostic_top100(useful,'no_market_final_score')
useful['no_market_selected'] = useful.merchant_abn.isin(no_market_ids)
useful['no_market_minus_baseline_score'] = useful.no_market_final_score-useful.final_score
useful['rank_improvement_without_market'] = useful.final_rank-useful.no_market_final_rank
# Renormalization also increases the four other business contributions; this is not just -0.09*MarketScore.
useful['removed_market_points'] = -.09*useful.market_score
useful['redistributed_business_points'] = .9*sum(useful[c]*(no_market_weights[c]-SCENARIOS['balanced'][c]) for c in no_market_weights)
np.testing.assert_allclose(useful.no_market_minus_baseline_score,useful.removed_market_points+useful.redistributed_business_points,atol=1e-12)
useful['market_switch_direction'] = np.select(
    [~useful.current_selected & useful.no_market_selected, useful.current_selected & ~useful.no_market_selected],
    ['Switch in without Market','Switch out without Market'],default='Unchanged membership')
market_cols = ['merchant_abn','merchant_name','market_switch_direction','current_selected','no_market_selected',
    'final_rank','no_market_final_rank','rank_improvement_without_market','seifa_irsad_national_decile','market_score',
    'final_score','no_market_final_score','no_market_minus_baseline_score','removed_market_points','redistributed_business_points',
    'regional_data_coverage_rate_all_sources_by_count','knn_confidence_score','knn_out_of_distribution']
market_switchers = useful.loc[useful.current_selected.ne(useful.no_market_selected),market_cols].sort_values('final_rank')
market_switchers.to_csv(RESULTS_DIR/'market_score_ablation_switchers.csv',index=False)
market_boundary_mask = useful.final_rank.between(80,120) | useful.no_market_final_rank.between(80,120) | useful.current_selected.ne(useful.no_market_selected)
market_boundary = useful.loc[market_boundary_mask,market_cols].sort_values('final_rank')
market_boundary.to_csv(RESULTS_DIR/'market_score_ablation_boundary.csv',index=False)
market_summary = pd.DataFrame([dict(comparison='Baseline vs No-Market',top100_overlap=len(baseline_ids&no_market_ids),
    switch_in_count=len(no_market_ids-baseline_ids),switch_out_count=len(baseline_ids-no_market_ids),
    spearman_all_eligible=useful.final_score.rank().corr(useful.no_market_final_score.rank()),
    baseline_top100_abs_rank_change_median=useful.loc[useful.current_selected,'rank_improvement_without_market'].abs().median(),
    baseline_top100_abs_rank_change_max=useful.loc[useful.current_selected,'rank_improvement_without_market'].abs().max(),
    eligible_irsad_p25=useful.seifa_irsad_national_decile.quantile(.25),
    eligible_irsad_median=useful.seifa_irsad_national_decile.median(),
    eligible_irsad_p75=useful.seifa_irsad_national_decile.quantile(.75),
    baseline_market_business_weight=.10,no_market_business_weight=0,
    no_market_value_weight=no_market_weights['value_score'],no_market_customer_weight=no_market_weights['customer_score'],
    no_market_growth_weight=no_market_weights['growth_score'],no_market_stability_weight=no_market_weights['stability_score'])])
market_summary.to_csv(RESULTS_DIR/'market_score_ablation_summary.csv',index=False)
display(market_summary.T)
display(market_switchers)
# Pick the closest raw-IRSAD opposite-direction pair, explicitly illustrative rather than representative.
in_rows = market_switchers.loc[market_switchers.no_market_selected]
out_rows = market_switchers.loc[market_switchers.current_selected]
pair_rows=[]
if len(in_rows) and len(out_rows):
    pairs = in_rows.merge(out_rows,how='cross',suffixes=('_in','_out'))
    pairs['raw_gap_abs']=(pairs.seifa_irsad_national_decile_in-pairs.seifa_irsad_national_decile_out).abs()
    closest=pairs.sort_values(['raw_gap_abs','merchant_abn_in','merchant_abn_out']).iloc[0]
    pair_rows=[dict(switch_in_abn=closest.merchant_abn_in,switch_in_name=closest.merchant_name_in,
        switch_out_abn=closest.merchant_abn_out,switch_out_name=closest.merchant_name_out,
        raw_irsad_in=closest.seifa_irsad_national_decile_in,raw_irsad_out=closest.seifa_irsad_national_decile_out,
        raw_irsad_gap_out_minus_in=closest.seifa_irsad_national_decile_out-closest.seifa_irsad_national_decile_in,
        market_score_gap_out_minus_in=closest.market_score_out-closest.market_score_in,
        baseline_market_points_gap_out_minus_in=.09*(closest.market_score_out-closest.market_score_in),
        baseline_final_score_gap_out_minus_in=closest.final_score_out-closest.final_score_in,
        no_market_final_score_gap_out_minus_in=closest.no_market_final_score_out-closest.no_market_final_score_in)]
market_pair=pd.DataFrame(pair_rows)
market_pair.to_csv(RESULTS_DIR/'market_score_ablation_irsad_pair.csv',index=False)
display(market_pair)
display(Markdown(f"**Selection effect:** removing and redistributing MarketScore changes {len(no_market_ids-baseline_ids)} slots "
    f"despite all-pool rank correlation {market_summary.iloc[0].spearman_all_eligible:.4f}. "
    "The paired raw/percentile differences above expose the scale conversion, while boundary tables show actual rank movements. "
    "Movement also includes redistribution to other business dimensions, so do not attribute the entire movement to raw IRSAD alone. "
    "Customer-area context is not merchant location or individual income; this sensitivity warrants discussing the substantive rationale and calibration of the contextual preference, not changing it automatically."))

,0
comparison,Baseline vs No-Market
top100_overlap,82
switch_in_count,18
switch_out_count,18
spearman_all_eligible,0.991034
baseline_top100_abs_rank_change_median,16.0
baseline_top100_abs_rank_change_max,89
eligible_irsad_p25,5.383429
eligible_irsad_median,5.474345
eligible_irsad_p75,5.55801


,merchant_abn,merchant_name,market_switch_direction,current_selected,no_market_selected,final_rank,no_market_final_rank,rank_improvement_without_market,seifa_irsad_national_decile,market_score,final_score,no_market_final_score,no_market_minus_baseline_score,removed_market_points,redistributed_business_points,regional_data_coverage_rate_all_sources_by_count,knn_confidence_score,knn_out_of_distribution
2165,57699602880,Tellus Id Institute,Switch out without Market,True,False,53,142,-89,5.571531,76.775956,82.398140,83.269573,0.871433,-6.909836,7.781269,0.802175,0.032787,True
2894,74063123996,Nec Eleifend Non Ltd,Switch out without Market,True,False,57,121,-64,5.527326,68.479881,82.313477,83.879370,1.565893,-6.163189,7.729083,0.806271,0.032787,True
3450,86710922099,Ac Urna Consulting,Switch out without Market,True,False,58,135,-77,5.536380,70.839543,82.294283,83.464429,1.170147,-6.375559,7.545706,0.812469,0.180328,False
1458,41663117354,Litora Torquent Industries,Switch out without Market,True,False,68,138,-70,5.534081,70.094386,82.000728,83.369852,1.369124,-6.308495,7.677619,0.809156,0.065574,False
2345,62224020443,Hendrerit A Corporation,Switch out without Market,True,False,69,119,-50,5.513789,64.704421,81.985873,83.914512,1.928640,-5.823398,7.752038,0.805861,0.377049,False
2461,64732735902,Imperdiet Non Vestibulum Institute,Switch out without Market,True,False,76,123,-47,5.506187,62.071535,81.817109,83.872264,2.055155,-5.586438,7.641593,0.805521,0.032787,True
1621,45380641195,Nisl Arcu Iaculis Incorporated,Switch out without Market,True,False,77,154,-77,5.543837,72.429210,81.791220,82.981325,1.190105,-6.518629,7.708734,0.801404,0.032787,True
56,11237511112,Magna Institute,Switch out without Market,True,False,80,116,-36,5.504226,61.500248,81.775107,84.015470,2.240362,-5.535022,7.775385,0.805849,0.459016,False
3961,98416475066,Lacinia Orci PC,Switch out without Market,True,False,81,120,-39,5.503333,61.276701,81.764623,83.909818,2.145195,-5.514903,7.660098,0.800620,0.032787,True
9,10323485998,Nunc Inc.,Switch out without Market,True,False,85,114,-29,5.495281,58.395430,81.689394,84.049527,2.360133,-5.255589,7.615722,0.808612,0.032787,True


,switch_in_abn,switch_in_name,switch_out_abn,switch_out_name,raw_irsad_in,raw_irsad_out,raw_irsad_gap_out_minus_in,market_score_gap_out_minus_in,baseline_market_points_gap_out_minus_in,baseline_final_score_gap_out_minus_in,no_market_final_score_gap_out_minus_in
0,39649557865,Arcu Morbi Institute,77590625261,Sed Diam Foundation,5.473365,5.489136,0.01577,6.607054,0.594635,0.207201,-0.491384


**Selection effect:** removing and redistributing MarketScore changes 18 slots despite all-pool rank correlation 0.9910. The paired raw/percentile differences above expose the scale conversion, while boundary tables show actual rank movements. Movement also includes redistribution to other business dimensions, so do not attribute the entire movement to raw IRSAD alone. Customer-area context is not merchant location or individual income; this sensitivity warrants discussing the substantive rationale and calibration of the contextual preference, not changing it automatically.

In [44]:
# Diagnostics must not mutate the formal ranking or overwrite any official recommendation list.
pd.testing.assert_frame_equal(final_df,locked_snapshot)
np.testing.assert_allclose(useful.final_score,.9*useful.balanced_score+.1*useful.risk_safety_score)
assert baseline_ids == set(top100_final.merchant_abn)
assert useful.merchant_abn.is_unique and len(useful)==4026
assert all(len(ids)==100 for ids in final_scenario_sets.values())
assert len(no_market_ids)==len(value_ids)==100
assert (review_list.ood_and_selection_sensitive <= review_list.selection_sensitive).all()
print('Useful diagnostics passed; locked scores and official selection remain unchanged.')

Useful diagnostics passed; locked scores and official selection remain unchanged.
